FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -65,4 +81,198 @@ object PgpMsg {
   )
 
   private val maxTagNumber = 20
+
+  enum class DecryptionErrorType {
+    KEY_MISMATCH,
+
+    // skip, effective for symmetric encryption
+    // USE_PASSWORD,
+
+    WRONG_PASSWORD,
+    NO_MDC,
+    BAD_MDC,
+
+    // a passphrase was not provided for the key which requires it,
+    // not sure how to detect this in the PGPainless.
+    NEED_PASSPHRASE,
+
+    FORMAT,
+    OTHER
+  }
+
+  data class DecryptionError(
+      val type: DecryptionErrorType,
+      val message: String,
+      val cause: Throwable? = null
+  )
+
+  data class DecryptionResult(
+      // provided if decryption was successful
+      val content: ByteArrayOutputStream?,
+
+      // true if message was encrypted.
+      // Alternatively false (because it could have also been plaintext signed,
+      // or wrapped in PGP armor as plaintext packet without encrypting,
+      val isEncrypted: Boolean,
+
+      // pgp messages may include original filename in them
+      val filename: String?,
+
+      // todo later - signature verification not supported on Android yet
+      val signature: Any?,
+
+      // provided if error happens
+      val error: DecryptionError?
+  ) {
+    companion object {
+      fun withError(
+          type: DecryptionErrorType,
+          message: String,
+          cause: Throwable? = null
+      ): DecryptionResult {
+        return DecryptionResult(
+            content = null,
+            isEncrypted = false,
+            filename = null,
+            signature = null,
+            error = DecryptionError(type, message, cause)
+        )
+      }
+
+      fun withCleartext(cleartext: ByteArrayOutputStream, signature: Any?): DecryptionResult {
+        return DecryptionResult(
+            content = cleartext,
+            isEncrypted = false,
+            filename = null,
+            signature = signature,
+            error = null
+        )
+      }
+
+      fun withDecrypted(content: ByteArrayOutputStream, filename: String?): DecryptionResult {
+        return DecryptionResult(
+            content = content,
+            isEncrypted = false,
+            filename = null,
+            signature = null,
+            error = null
+        )
+      }
+    }
+  }
+
+  @Suppress("ArrayInDataClass")
+  data class SecretKeyInfo(
+      val armored: String? = null,
+      val keyRing: PGPSecretKeyRing = readArmored(armored!!),
+      val passphrase: CharArray? = null
+  ) {
+    companion object {
+      private fun readArmored(armored: String): PGPSecretKeyRing {
+        return JcaPGPSecretKeyRingCollection(
+            ArmoredInputStream(armored.toByteArray().inputStream())
+        ).keyRings.asSequence().first()
+      }
+    }
+  }
+
+  fun decrypt(
+      data: ByteArray,
+      pgpSecretKeyRing: PGPSecretKeyRing, // for decryption
+      passphrase: Passphrase?, // for decryption of secret keys
```

Comment:
Each key may be protected by a different pass phrase. I think we could use some type like `KeyWithOptionalPassPhrase[]` when one key matches one possible pp.

Reply:
Each key or each key in the keyring? PGPainless does not handle such case as somethings standard. If that is the case, maybe it is worth to implement in the PGPainless. Also as you can see, we only support only one keyring now, so such data structure may be postponed.

Reply:
Sorry, I'm not too familiar with Java PGP terminology. Here is how I'm using the words:

key packet - An OpenPGP key packet like primary key packet or subkey packet
OpenPGP key - a collection of usually one master key packet and some number of sub-packets

I don't know what a keyring is, but presumably it's a collection of OpenPGP keys.

Each OpenPGP key may have a different pass phrase. Whether these are represented during decryption as one key ring, or several key rings, or several OpenPGP keys differently is an implementation detail (which we may need to resolve).

Technically, each key packet within OpenPGP key may have a different pass phrase, but we don't support this scenario (and check for it during setup), so you will not encounter such keys on Android.

Now, for this actual issue, you may use a temporary solution as in https://github.com/FlowCrypt/flowcrypt-android/pull/1216#discussion_r628708349 (decrypt all yourself first) and then later we include this functionality in PGPainless, so that we don't have to decrypt keys and packets that are not used for a particular message.

Reply:
@tomholub BouncyCastle `PGPKeyRing` (and its descendants `PGPPublicKeyRing` and `PGPSecretKeyRing`, which you can see here in the code) matches to what you call "OpenPGP key". Keeping this in mind, we can say that current decryption routine supports only one "OpenPGP key". Should we pass there multiple ones? And then decrypt them and then pass decrypted ones to the PGPainess?

Reply:
Thank you for explanation. While most users will have one OpenPGP key, they may also have more than one. Therefore we'll have to support supplying more than one OpenPGP key (or PGPSecretKeyRing) into the decrypt function.

As you say, we would decrypt them all first, and then supply into PGPainless decrypted.

In the future, we have to improve PGPainless to:
 - accept more than one PGPSecretKeyRing (if it doesn't already)
 - allow for different PGPSecretKeyRing to be unlocked by different pass phrase

But that's an optimization, after we drop NodeJS on Android.

> Should we pass there multiple ones? And then decrypt them and then pass decrypted ones to the PGPainess?

:+1:

Reply:
> accept more than one PGPSecretKeyRing (if it doesn't already)

I think PGPainless already did that - correct me if I'm wrong. If it already supports passing several, then this will be easier. If not, they you'd need to call it in a loop, key by key..